plugin/trino-password-authenticators/src/main/java/io/trino/plugin/password/ldap/LdapGroupProvider.java

```diff
@@ -0,0 +1,324 @@
+            String searchFilter,
+            String bindDistinguishedName,
+            String bindPassword)
+            throws NamingException
+    {
+        Set<String> userDistinguishedNames = lookupUserDistinguishedNames(userBaseDistinguishedName,
+                searchFilter, bindDistinguishedName, bindPassword);
+        if (userDistinguishedNames.isEmpty()) {
+            if (allowUserNotExist) {
+                return "";
+            }
+            else {
+                String message = format("User filter [%s] failed for baseDN [%s]", searchFilter, userBaseDistinguishedName);
+                log.debug("%s", message);
+                throw new AccessDeniedException(message);
+            }
+        }
+        if (userDistinguishedNames.size() > 1) {
+            String message = format("Multiple group membership results for user [%s]: %s", userBaseDistinguishedName,
+                    userDistinguishedNames);
+            log.debug("%s", message);
+            throw new AccessDeniedException(message);
+        }
+        return getOnlyElement(userDistinguishedNames);
+    }
+
+    static String extractGroupName(
+            String source,
+            String groupNameAttribute,
+            String groupUserMembershipAttribute)
+            throws NamingException
+    {
+        LdapName ldn;
+        try {
+            ldn = new LdapName(source);
+        }
+        catch (InvalidNameException e) {
+            if (!source.isEmpty() && groupUserMembershipAttribute != null && groupUserMembershipAttribute.equalsIgnoreCase("memberOf")) {
+                return source;
+            }
+            throw e;
+        }
+        List<Rdn> rdns = ldn.getRdns();
+        if (rdns.isEmpty()) {
+            throw new NamingException("ExtractGroupName - DN is empty");
+        }
+        Rdn rdn = rdns.get(rdns.size() - 1);
+        if (rdn.getType().equalsIgnoreCase(groupNameAttribute)) {
+            return (String) rdn.getValue();
+        }
+        throw new NamingException("ExtractGroupName - Unable to find a group name using a " +
+                "group-name-attribute [" + groupNameAttribute + "]. The DN is [" + source + "]");
+    }
+
+    static Set<String> extractGroupNameFromUser(
+            Set<String> userGroups,
+            String groupNameAttribute,
+            String groupUserMembershipAttribute)
+    {
+        Set<String> groupSet = userGroups.stream().map(source -> {
+            log.debug("extracting group from %s", source);
+            try {
+                return extractGroupName(source, groupNameAttribute, groupUserMembershipAttribute);
+            }
+            catch (NamingException e) {
+                log.debug("Cannot extract group [%s] [%s] [%s]", groupNameAttribute, groupUserMembershipAttribute, source);
+            }
+            return null;
+        }).filter(Objects::nonNull).collect(Collectors.toUnmodifiableSet());
```

Comment:
Any specific for not using guava `ImmutableSet` 
Ref : https://github.com/trinodb/trino/blob/master/.github/DEVELOPMENT.md#prefer-guava-immutable-collections

Reply:
yeah, i did see this.  Is there a reason that core Java Collectors.toUnmodifiableSet() is not OK here? We do not require ordered Sets - and if we did, would use a different container to explicitly make that behavior so. I can change it but i can see several other uses if this in the code base outside ldap.

Reply:
The main reason is deterministic iteration which could help us during debugging and to have consistency across the code base.  


> i can see several other uses of this in the code base outside ldap.

The usage is specific to iceberg connector I guess - other part of the codebase still use `ImmutableSet` instead of `UnmodifiableSet`.

Reply:
`.filter` and `.collect` can be moved to a new line